### ISSUE TITLE
ServiceConnectionError Exception

### ISSUE THREAD

**Issue by nodatasheet** [Animenosekai/translate#57]
Hi guys!

First of all, awesome project, many thanks for it!

Now to my issue: 
I can not figure out a proper way of checking whether `service_url` is reachable by `translatepy` or not (some kind of ServiceConnectionError Exception).

If i just use this code when I'm offline:
```python
from translatepy.translators.google import GoogleTranslate

server_url = 'translate.google.com'

translator = GoogleTranslate(service_url=server_url)
translator.translate('Текст 15', 'en').result

```

I get error: 
```batch
translatepy\translators\base.py", line 110, in translate
    source_language, translation = self._translate(text, dest_code, source_code)
TypeError: cannot unpack non-iterable NoneType object
```

Which is not a clear reason of the translation failure. 

I tried to check it myself prior to translation using `requests`:
```python
import requests
from translatepy.translators.google import GoogleTranslate


def is_online(url):
    try:
        response = requests.head(url)
    except Exception as e:
        return f'{url} not reachable: {str(e)}'
    else:
        stat_code = response.status_code
        if stat_code == 200:
            return True
        else:
            return f'{url} not online. HTTP response code: {stat_code}'


server_url = 'translate.google.com'

print(is_online(f'https://{server_url}'))

translator = GoogleTranslate(service_url=server_url)
translator.clean_cache()
translator.translate('Текст 16', 'en').result
```

But it confuses me even more, as sometimes it gives opposite results. 
Like, url is not reachable but translation is ok (or vice versa):
```batch
https://translate.google.com not reachable: HTTPSConnectionPool(host='translate.google.com', port=443): Max retries exceeded with url: / (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x000001D488901930>, 'Connection to translate.google.com timed out. (connect timeout=None)'))
'Text 16'
```

This happens when I'm turning on/off VPN instead of shutting down the network. 

What I need is a clear error that service's url could not be reached by `translatepy`. 
Any suggestions on this issue?

Thanks in advance!

**Comment by Animenosekai** [comment]
About this error now :

> But it confuses me even more, as sometimes it gives opposite results.
> Like, url is not reachable but translation is ok (or vice versa)

We indeed see in your example that the server is not reachable but translatepy gives back a translated result 🤔 (wow awesome)

```python
https://translate.google.com not reachable: HTTPSConnectionPool(host='translate.google.com', port=443): Max retries exceeded with url: / (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x000001D488901930>, 'Connection to translate.google.com timed out. (connect timeout=None)'))
'Text 16'
```

What I guess is happening here is that you translated the text in the past before turning off your network connection, which kept it in cache.

```python
# translatepy/utils/request.py#L179
        _cache_key = str(url) + str(kwargs)
        if _cache_key in self.GETCACHE and time() - self.GETCACHE[_cache_key]["timestamp"] < self.cache_duration:
            return self.GETCACHE[_cache_key]["response"]
```

Even with your network connection turned off, the request handler eventually hit the cache and returned you the cached response.

**Comment by nodatasheet** [comment]
@Animenosekai Thanks for quick reply!

I think I can live with this issue for now ☺ 

I'll do `raise_for_status()` check before the translation. So it won't leave me with unpredictable result.
___
> What I guess is happening here is that you translated the text in the past before turning off your network connection, which kept it in cache.

Could that happen even if I cleaned the cache (note `translator.clean_cache()` in previous example)? 

I think it's all about my location (China).

this code:
```python
import requests
from translatepy.translators.google import GoogleTranslate

def is_online(url):
    try:
        response = requests.get(url)
        response.raise_for_status()
        return response.status_code
    except Exception as error:
        return error

server_url = 'translate.google.com'

check_online = is_online(f"https://{server_url}")
print(check_online)

translator = GoogleTranslate(service_url=server_url)
print(translator.translate('Совсем-совсем новый текст', 'en').as_json)
translator.clean_cache()

```

1. When online but vpn is off:
```python
HTTPSConnectionPool(host='translate.google.com', port=443): Max retries exceeded with url: / (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x0000021F089E0340>, 'Connection to translate.google.com timed out. (connect timeout=None)'))
<bound method TranslationResult.as_json of TranslationResult(service=Google, source=Совсем-совсем новый текст, source_language=rus, destination_language=eng, result=Completely new text)>
```

2. When completely offline:
```python
HTTPSConnectionPool(host='translate.google.com', port=443): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000001EB74AE8700>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))
Traceback (most recent call last):
  File "...\WIP-2.py", line 18, in <module>
    print(translator.translate('Совсем-совсем новый текст', 'en').as_json)
  File "...\Python310\lib\site-packages\translatepy\translators\base.py", line 110, in translate
    source_language, translation = self._translate(text, dest_code, source_code)
TypeError: cannot unpack non-iterable NoneType object
```

Online, without vpn and changing server to `translate.google.cn`:
```python
200
<bound method TranslationResult.as_json of TranslationResult(service=Google, source=Совсем-совсем новый текст, source_language=rus, destination_language=eng, result=Very, very new text)>
```

My guess is that even if `request` in check could not get to google.com, translator still can. I assume mystery solved. 

Many thanks for helping!

**Comment by Animenosekai** [comment]
> Could that happen even if I cleaned the cache (note translator.clean_cache() in previous example)?

Yes because this cleans the translator's cache (aka the one with the successful results)

You can give your own `Request` class

> Example
```python
from translatepy.utils.request import Request
from translatepy.translators.google import GoogleTranslate

request_handler = Request()
translator = GoogleTranslate(request=request_handler)
result = translator.translate("Hello world", "japanese", "english")
print(result.as_json())
translator.clean_cache()
request_handler.GETCACHE.clear()
```

**Comment by Animenosekai** [comment]
> Online, without vpn and changing server to translate.google.cn

Yes, indeed, the `service_url` parameter was made mostly because some people need to access the `.cn` server.

**Comment by nodatasheet** [comment]
> You can give your own `Request` class


Thanks for the hint. But unfortunately it gave me same results.

```python
from translatepy.utils.request import Request
from translatepy.translators.google import GoogleTranslate


def trpy_request_status(request_handler, url):
    try:
        response = request_handler.get(url)
        response.raise_for_status()
        return response.status_code
    except Exception as error:
        return error


server_name = "translate.google.com"
server_url = f"https://{server_name}"

request_handler = Request()
print(trpy_request_status(request_handler, server_url))
translator = GoogleTranslate(request=request_handler, service_url=server_name)
result = translator.translate("Hello world", "jpn", "en")
print(result.as_json())
translator.clean_cache()
request_handler.GETCACHE.clear()
```
```
HTTPSConnectionPool(host='translate.google.com', port=443): 
    Max retries exceeded with url: 
        / (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x00000254E89F0CA0>, 
            'Connection to translate.google.com timed out. (connect timeout=None)'))

{"success": true, 
 "service": "Google",
 "source": "Hello world", 
 "sourceLanguage": "eng",
 "destinationLanguage": "jpn",
 "result": "\u3053\u3093\u306b\u3061\u306f\u4e16\u754c"}
```
So, I'll stick to the logic that if request did not reach it, I won't pass it. Even if it can translate afterwards, it is still too slow.

**Comment by Animenosekai** [comment]
> But unfortunately it gave me same results.

The `request_handler.GETCACHE.clear()` is what clears the "GET" requests caches so you need to test after this call if you don't want to use the cache (even if it's better if it caches it isn't it?)

Also, you should call the `translator.clean_caches()` method because if it succeeded before the translation is in this cache too

**Comment by nodatasheet** [comment]
For purity I was changing the source text every time during the tests 😂. So cleaning would not affect the results anyway, right?

> even if it's better if it caches it isn't it?

For sure I agree it is much better with it. But I'm not familiar with caching topic and wasn't clear about how big it can grow, also where it is stored. So for myself I keep it on, for public I keep it off by default with possibility of turning on. 

Maybe it is not the topic of this thread, but could you please clarify that? In [lru_chacher](translatepy/utils/lru_cacher.py) I see the max size is 1024(kb?). So, it can not grow bigger?

**Comment by Animenosekai** [comment]
> Maybe it is not the topic of this thread

Yup, it's not really the topic of this thread, but we can make something like a “Discussion” about it (it is enabled in this repository)

> I see the max size is 1024(kb?)

I also thought about that at first (because it was implemented by one of the greatest contributors to this repo @ZhymabekRoman) but looking at it, it seems to be the number of elements which can be in the object.

We could make another issue to decide whether to change to a bytes-sized based approach or not.

**Comment by ZhymabekRoman** [comment]
> We could make another issue to decide whether to change to a bytes-sized based approach or not.

I think this is a better idea than what is implemented now. This would allow a rational use of RAM. I once thought to implement weakref LRU cache, but I got some messy code.

**Comment by nodatasheet** [comment]
> I also thought about that at first (because it was implemented by one of the greatest contributors to this repo @ZhymabekRoman) but looking at it, it seems to be the number of elements which can be in the object.

 I just read a bit more about `functools.lru_cache` and I think I can skip the clearing by default in my code. 
1024 calls is not such a big thing in my case. Even with 5000 characters per call it is how many.. ~15Mb? Not a big deal 😂
Won't even expose it out of the hood actually.
Thanks for clarification 👍

**Comment by Animenosekai** [comment]
Closing this as the main issue should be solved with 0e789d340fce7863e88c435b2128ad16f1aae699, we will continue this in other issues like #58
